pkg/operator/ceph/file/filesystem.go

```diff
@@ -55,7 +55,7 @@ func createFilesystem(
 		return err
 	}
 
```

Comment:
I assume it shouldn't be supported for users to specify both of these. If so, we should also include a check to make sure that the user has not specified both DataPools and DataPoolsMapped. Does that make sense to you @travisn ?

Reply:
There are a number of important design points that come to mind with this feature. What about the following?
- For backward compatibility, we should still support the `dataPools`. Let's keep it as a fully supported feature rather than deprecating it. Perhaps we eventually deprecate it, but with some unit tests it seems like it should be very simple to support. 
- Let's only allow either `dataPools` or `namedDataPools`. No need to support both in the same filesystem CR. The operator would fail the reconcile if both are specified.
- The example manifests and docs could all be updated to use namedDataPools by default.
- Ordering is important for adding the pools to cephfs. As seen in #8452, to support EC, a replica data pool must be created first, then the EC pool. So instead of a map we need to keep an ordered list. For the types.go, this might look like the following where the PoolSpec is an inline property:
```go
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []PoolSpec `json:"dataPools"`
	NamedDataPools []NamedPoolSpec `json:"namedDataPools"`
...
}
```

- For the implementation, before the ceph pools are created, we would translate all the dataPools into namedDataPools with the default pool names. Then all the pools can be processed the same. No need for a separate implementation in the operator code for creating the ceph pools.
- If users desired to update their CRs to the named pools, they would simply need to rename `dataPools` to `namedDataPools` and set the `name` of each pool to the name that was being generated previously by rook.

Reply:
In that case we may follow way:

```
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
        Default bool `json:"default"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []PoolSpec `json:"dataPools"`
	NamedDataPools []NamedPoolSpec `json:"namedDataPools"`
...
}
```

In that case we dont need an ordered list, because ordering seems to me is a hardcode and workaround. Before filesystem is going to be created check that default pool is present and it is the only one. For backward compatibility, since right now 1 element of array is assuming to be default we dont have any problems with transition to a new structure.

@travisn  @BlaineEXE what do you think?

Reply:
Agreed, the designation of the first/default pool is the only thing cephfs cares about. The ordering of the additional data pools is unimportant to ceph. 

I still like the idea of keeping the ordered list. There are several advantages:
- No need for a new property to designate the default
- Follows the same pattern we've already been using, we're just adding a `name` property and renaming it to `namedDataPools`. 
- Very easy to convert existing filesystem CRs to the new approach.

Reply:
Actually, we should be able to get away with not renaming it to `namedDataPools`, we should just be able to add the `name` to the existing pools. Then if the name is empty, we generate the name.

```go
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []NamedPoolSpec `json:"dataPools"`
...
}
```

Reply:
After the initial creation of the filesystem, is there a way to change the primary pool? For example, if you remove a named pool, it wouldn't affect the other named pools being removed. Or what is the issue with pool removal?

If we use the inline flag, I believe the compatibility would be fine, it would just be handled by the json deserialization inside the operator.
```go
type NamedPoolSpec struct {
	Name     string `json:"name,omitempty"`
	PoolSpec `json:",inline"`
}
```

But another thought I'm having is that we could just add the `name` property to the PoolSpec. This would allow the user to specify pool names on the cephfs pools to override the default, and with #9209 for configuring the built-in ceph pool. It just wouldn't make sense in some areas like the rgw pool names, which I don't believe are configurable, so maybe that approach is less preferred.

Reply:
Ok, so if we are assuming that first element of `dataPools` array is always coming as default one (on creation) and we dont manage pool removal for cephfs, let's go your way.

Another reason for specifying `default` property is adding validation to avoid issue https://github.com/rook/rook/pull/8452 - give to a user an explicit error message what is going wrong without double checking docs. But since user is allowed to drop from spec any pool without its actual removal, then it will not work correctly. In the same time, we can add such validation on creation. Do you agree? I will post a comment in related issue.

I'll post update as discussed above, with auto transition to a new `NamedPoolSpec` structure, all pools present in current `dataPools` will be named by default pool name + its index.

Reply:
Looks like there is a request on #8452 to not prevent the creation in that case. Instead of blocking the case, logging a warning seems sufficient.

> I'll post update as discussed above, with auto transition to a new NamedPoolSpec structure, all pools present in current dataPools will be named by default pool name + its index.

Agreed on this approach, thanks.